### DISCUSSION TITLE
Generating and using a JWT in a composite action

### DISCUSSION THREAD

**Discussion by TheYoctoJester** [community/community#26309] [Actions]
<p>Howdy!</p>
<p>I am writing a composite action that first retrieves a JWT through a login-type api call, and then uses it as authentication bearer for <code>curl</code>. This works fine if done as a script in a Dockerfile. However in the composite action, there seems to be some expansion problem of the JWT. The line in question is <a href="https://github.com/TheYoctoJester/mender-push-artifact/blob/69e1e612bb0c0cf62e65f1acafc4f643e7745842/action.yml#L35" rel="noopener nofollow ugc">here</a>, and I would be glad if somebody can tell if my line of thinking is right or has some tips to tackle this.</p>
<p>Greetz,<br>
Josef</p>

**Comment by airtower-luna** [comment]
<p>Where is the <code>JWT</code> environment variable created? The only place I see it used otherwise is here:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/TheYoctoJester/mender-push-artifact/blob/69e1e612bb0c0cf62e65f1acafc4f643e7745842/action.yml#L22">
  <header class="source">

      <a href="https://github.com/TheYoctoJester/mender-push-artifact/blob/69e1e612bb0c0cf62e65f1acafc4f643e7745842/action.yml#L22" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/TheYoctoJester/mender-push-artifact/blob/69e1e612bb0c0cf62e65f1acafc4f643e7745842/action.yml#L22" target="_blank" rel="noopener">TheYoctoJester/mender-push-artifact/blob/69e1e612bb0c0cf62e65f1acafc4f643e7745842/action.yml#L22</a></h4>



    <pre class="onebox">      <code class="lang-yml">
        <ol class="start lines" start="12" style="counter-reset: li-counter 11 ;">
            <li>    description: 'URI for the hosted Mender backend account to be used'</li>
            <li>    required: false</li>
            <li>    default: 'https://hosted.mender.io'</li>
            <li>  mender_artifact:</li>
            <li>    description: 'path of the artifact to be uploaded, relative to GITHUB_WORKSPACE'</li>
            <li>    required: true</li>
            <li>runs:</li>
            <li>  using: "composite"</li>
            <li>  steps:</li>
            <li>    - id: get_access_token</li>
            <li class="selected">      run: JWT=$(curl -s -X POST -u ${{ inputs.mender_user }}:${{ inputs.mender_password }} ${{ inputs.mender_uri }}/api/management/v1/useradm/auth/login)</li>
            <li>      shell: bash</li>
            <li>    - id: show_environment</li>
            <li>      run: env</li>
            <li>      shell: bash</li>
            <li>    - id: show_GITHUB_WORKSPACE</li>
            <li>      run: tree $GITHUB_WORKSPACE</li>
            <li>      shell: bash</li>
            <li>    - id: upload_artifact</li>
            <li>      run: |</li>
            <li>        curl -s -X POST ${{ inputs.mender_uri }}/api/management/v1/deployments/artifacts \</li>
        </ol>
      </code>
    </pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>That’s in a different step though, which runs in a separate shell, so its variables are long gone by the time <code>curl</code> runs. The safest way to handle it is probably to set the token as a step output in the <code>get_access_token</code> step (maybe mask it, too?), and then use that output in the <code>upload_artifact</code> step.</p>

**Comment by TheYoctoJester** [comment]
<p>Thanks, yeah… the variable wasn’t being handled correctly. Being rather new to GA, coming from classic shell scripting. It seems to work now with the assignment being written into $GITHUB_ENV, and retrieved with <code>${{ env.JWT }}</code>.</p>
<p>Just a small follow-up - how does this compare to the output strategy that you mentioned? Up-/downsides?</p>
<p>Thanks a lot!</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="TheYoctoJester" data-post="3" data-topic="255412">
<div class="title">
<div class="quote-controls"></div>
 TheYoctoJester:</div>
<blockquote>
<p>Just a small follow-up - how does this compare to the output strategy that you mentioned? Up-/downsides?</p>
</blockquote>
</aside>
<p>Writing to <code>GITHUB_ENV</code> will make the token available to <em>everything</em> running in all later steps, including outside the composite action. Considering you’re handling an authentication token I’d consider that a security risk.</p>
<p>Well, and there’s the potential that something in the calling workflow also defined a <code>JWT</code> environment variable before and expects it to be available later.</p>
